MatchErr is set to ErrNotFound if NotFoundHandler is used

roobre opened the pull request:
As discussed in #293, it should be useful if `RouteMatch.MatchErr` were set to an error if no match is found, even if some subrouters have a custom `NotFoundHandler`.

elithrar commented on mux_test.go:
```diff
@@ -1982,3 +1982,54 @@ func TestNoMatchMethodErrorHandler(t *testing.T) {
+	matched := r.Match(req, match)
+
+	if matched {
+		t.Error("Subrouter should not have matched that")
```
I would add some context - `”subrouter should not have matched: got %v”, route”` to aid debugging.

elithrar commented on mux_test.go:
```diff
@@ -1982,3 +1982,54 @@ func TestNoMatchMethodErrorHandler(t *testing.T) {
+	matched = s.Match(req, match)
+	// Now we should get a match
+	if !matched {
+		t.Error("Subrouter should have matched that")
```
Here too.

elithrar commented on mux_test.go:
```diff
@@ -1982,3 +1982,54 @@ func TestNoMatchMethodErrorHandler(t *testing.T) {
+
+	// Now we should get a match
+	if !matched {
+		t.Error("Subrouter should have matched that")
```
Here too.

roobre commented:
I've added the matching route where none is expected but one is got. I've also moved the last two test cases above the private functions, which I think is more appropriate.

The go fmt issue with travis is bugging me a bit, but afaik it's not my fault. It appeared when I merged upstream :/

elithrar commented:
Thanks for this. Let me see why gofmt is failing [only] on tip.

elithrar commented:
Rebase against master to pull in https://github.com/gorilla/mux/pull/312

On Fri, Nov 3, 2017 at 12:55 PM Roberto Santalla <notifications@github.com>
wrote:

> I've added the matching route where none is expected but one is got. I've
> also moved the last two test cases above the private functions, which I
> think is more appropriate.
>
> The go fmt issue with travis is bugging me a bit, but afaik it's not my
> fault. It appeared when I merged upstream :/
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/gorilla/mux/pull/311#issuecomment-341810035>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABIcFRthFq0H5NDllXQu0ATJaDGYz6Sks5sy2-ggaJpZM4QRfiH>
> .
>

kisielk commented on route.go:
```diff
@@ -72,7 +72,11 @@ func (r *Route) Match(req *http.Request, match *RouteMatch) bool {
 
-	match.MatchErr = nil
+	if match.MatchErr == ErrMethodMismatch {
+		// We found a non-mismatching route, clear MatchErr
```
non-mismatching is a bit of a double negative :) how about just "matching" so it's less confusing?

roobre replied:
Hm, I used that wording to make clear that the ErrMethod*Mismatch* found on the previous iteration no longer applies. But I guess I can reword it a bit.

I'll do it tomorrow tho, its 2AM here and I'm quite asleep.

kisielk commented:
LGTM other than comment

kisielk commented:
Thanks! I don't mean to be too nitpicky, but the code base is already pretty hard to follow so any little bit helps.

roobre commented:
@kisielk No prob! I'm not a native speaker, so I can make some grammar mistakes here and there. Also, I'd do the same on projects I mantain :P
